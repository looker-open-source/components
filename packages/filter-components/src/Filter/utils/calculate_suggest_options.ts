/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import difference from 'lodash/difference'
import find from 'lodash/find'
import isArray from 'lodash/isArray'
import map from 'lodash/map'
import take from 'lodash/take'
import union from 'lodash/union'
import nth from 'lodash/nth'
import pull from 'lodash/pull'

/*
Calculate the Suggestion Options based on max number of allowed
values as well as pre-existing values on the filter which should show
to allow de-selection
*/
export const calculateSuggestOptions = (filterTokenProps: any) => {
  const { max, options, value: values } = filterTokenProps

  // if values is not an arary or is not an array containing a string
  // return default
  if (!isArray(values) || !find(values, val => typeof val === 'string')) {
    return max ? take(options, max) : options
  }

  const existingValues =
    map(
      values,
      v => find(options, { value: v }) || { value: v, label: 'âˆ… - ' + v }
    ) || []

  // if there is a max limit, first use all existing values then options
  // if there is no max, union all
  if (max) {
    // if there are more values than max, only show existing values
    // else there is room to show other options
    if (existingValues.length >= max) {
      return take(existingValues, max)
    } else {
      const suggestionOptions: { value: string; label: string }[] = []
      for (
        let optionIndex = 0;
        optionIndex < max - existingValues.length;
        optionIndex++
      ) {
        const option = nth(options, optionIndex) as {
          value: string
          label: string
        }
        if (!option) {
          break
        }
        pull(existingValues, option)
        suggestionOptions.push(option)
      }
      return union(suggestionOptions, existingValues)
    }
  } else {
    const unfoundValues = difference(existingValues, options) || []
    return union(options, unfoundValues)
  }
}
