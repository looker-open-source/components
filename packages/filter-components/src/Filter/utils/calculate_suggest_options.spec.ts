/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { calculateSuggestOptions } from './calculate_suggest_options'

describe('Calculate Suggest Options', () => {
  const test_options = [
    { value: 'a', label: 'a' },
    { value: 'b', label: 'b' },
    { value: 'c', label: 'c' },
    { value: 'd', label: 'd' },
    { value: 'e', label: 'e' },
    { value: 'f', label: 'f' },
    { value: 'g', label: 'g' },
    { value: 'h', label: 'h' },
    { value: 'i', label: 'i' },
    { value: 'j', label: 'j' },
    { value: 'k', label: 'k' },
    { value: 'l', label: 'l' },
    { value: 'm', label: 'm' },
    { value: 'n', label: 'n' },
    { value: 'o', label: 'o' },
    { value: 'p', label: 'p' },
    { value: 'q', label: 'q' },
    { value: 'r', label: 'r' },
    { value: 's', label: 's' },
    { value: 't', label: 't' },
    { value: 'u', label: 'u' },
    { value: 'v', label: 'v' },
    { value: 'w', label: 'w' },
    { value: 'x', label: 'x' },
    { value: 'y', label: 'y' },
    { value: 'z', label: 'z' },
  ]

  it('Return existing values along with current options', () => {
    expect(
      calculateSuggestOptions({
        value: ['a', 'z'],
        options: test_options,
        max: 5,
      })
    ).toMatchSnapshot()
  })

  it('Return existing values along with current options 2', () => {
    expect(
      calculateSuggestOptions({
        value: ['a', 'z', 'zz', 'zba'],
        options: test_options,
        max: 5,
      })
    ).toMatchSnapshot()
  })

  it('Return limited existing values if over limit', () => {
    expect(
      calculateSuggestOptions({
        value: ['a', 'b', 'c', 'd', 'e', 'f', 'g'],
        options: ['x', 'y', 'z'],
        max: 5,
      })
    ).toMatchSnapshot()
  })
  it('Return only existing values if at limit', () => {
    expect(
      calculateSuggestOptions({
        value: ['a', 'b', 'c', 'd', 'e', 'f', 'g'],
        options: ['x', 'y', 'z'],
        max: 7,
      })
    ).toMatchSnapshot()
  })
  it('Return only options if no value', () => {
    expect(
      calculateSuggestOptions({
        value: [],
        options: test_options,
        max: 7,
      })
    ).toMatchSnapshot()
  })

  it('Return everything if no max', () => {
    expect(
      calculateSuggestOptions({
        value: ['aa', 'z', 'b', 'gg'],
        options: test_options,
        max: null,
      })
    ).toMatchSnapshot()
  })

  it('Return normally if no string', () => {
    expect(
      calculateSuggestOptions({
        value: [100, 200],
        options: [1, 2, 3, 4, 5],
        max: 3,
      })
    ).toMatchSnapshot()
  })
})
