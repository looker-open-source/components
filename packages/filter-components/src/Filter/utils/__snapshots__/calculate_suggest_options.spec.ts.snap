// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculate Suggest Options Return everything if no max 1`] = `
Array [
  Object {
    "label": "a",
    "value": "a",
  },
  Object {
    "label": "b",
    "value": "b",
  },
  Object {
    "label": "c",
    "value": "c",
  },
  Object {
    "label": "d",
    "value": "d",
  },
  Object {
    "label": "e",
    "value": "e",
  },
  Object {
    "label": "f",
    "value": "f",
  },
  Object {
    "label": "g",
    "value": "g",
  },
  Object {
    "label": "h",
    "value": "h",
  },
  Object {
    "label": "i",
    "value": "i",
  },
  Object {
    "label": "j",
    "value": "j",
  },
  Object {
    "label": "k",
    "value": "k",
  },
  Object {
    "label": "l",
    "value": "l",
  },
  Object {
    "label": "m",
    "value": "m",
  },
  Object {
    "label": "n",
    "value": "n",
  },
  Object {
    "label": "o",
    "value": "o",
  },
  Object {
    "label": "p",
    "value": "p",
  },
  Object {
    "label": "q",
    "value": "q",
  },
  Object {
    "label": "r",
    "value": "r",
  },
  Object {
    "label": "s",
    "value": "s",
  },
  Object {
    "label": "t",
    "value": "t",
  },
  Object {
    "label": "u",
    "value": "u",
  },
  Object {
    "label": "v",
    "value": "v",
  },
  Object {
    "label": "w",
    "value": "w",
  },
  Object {
    "label": "x",
    "value": "x",
  },
  Object {
    "label": "y",
    "value": "y",
  },
  Object {
    "label": "z",
    "value": "z",
  },
  Object {
    "label": "∅ - aa",
    "value": "aa",
  },
  Object {
    "label": "∅ - gg",
    "value": "gg",
  },
]
`;

exports[`Calculate Suggest Options Return existing values along with current options 1`] = `
Array [
  Object {
    "label": "a",
    "value": "a",
  },
  Object {
    "label": "b",
    "value": "b",
  },
  Object {
    "label": "c",
    "value": "c",
  },
  Object {
    "label": "d",
    "value": "d",
  },
  Object {
    "label": "z",
    "value": "z",
  },
]
`;

exports[`Calculate Suggest Options Return existing values along with current options 2 1`] = `
Array [
  Object {
    "label": "a",
    "value": "a",
  },
  Object {
    "label": "b",
    "value": "b",
  },
  Object {
    "label": "z",
    "value": "z",
  },
  Object {
    "label": "∅ - zz",
    "value": "zz",
  },
  Object {
    "label": "∅ - zba",
    "value": "zba",
  },
]
`;

exports[`Calculate Suggest Options Return limited existing values if over limit 1`] = `
Array [
  Object {
    "label": "∅ - a",
    "value": "a",
  },
  Object {
    "label": "∅ - b",
    "value": "b",
  },
  Object {
    "label": "∅ - c",
    "value": "c",
  },
  Object {
    "label": "∅ - d",
    "value": "d",
  },
  Object {
    "label": "∅ - e",
    "value": "e",
  },
]
`;

exports[`Calculate Suggest Options Return normally if no string 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`Calculate Suggest Options Return only existing values if at limit 1`] = `
Array [
  Object {
    "label": "∅ - a",
    "value": "a",
  },
  Object {
    "label": "∅ - b",
    "value": "b",
  },
  Object {
    "label": "∅ - c",
    "value": "c",
  },
  Object {
    "label": "∅ - d",
    "value": "d",
  },
  Object {
    "label": "∅ - e",
    "value": "e",
  },
  Object {
    "label": "∅ - f",
    "value": "f",
  },
  Object {
    "label": "∅ - g",
    "value": "g",
  },
]
`;

exports[`Calculate Suggest Options Return only options if no value 1`] = `
Array [
  Object {
    "label": "a",
    "value": "a",
  },
  Object {
    "label": "b",
    "value": "b",
  },
  Object {
    "label": "c",
    "value": "c",
  },
  Object {
    "label": "d",
    "value": "d",
  },
  Object {
    "label": "e",
    "value": "e",
  },
  Object {
    "label": "f",
    "value": "f",
  },
  Object {
    "label": "g",
    "value": "g",
  },
]
`;
